Add hashed and private read-write set to proto

cendhu opened the pull request:
This CR adds protobuf struct for hashed read/write set and private
read/write set.

To the existing KVRWSet, i have added a field to store hashed read write
set of all collections. For a private read, we can store the hash in key
field but keep the version as is. For private writes, we can store hash
of key and hash of value but store the IsDelete as is. Hence, reused
existing KVRead and KVWrite.

For the range query info, it does not make sense to store the hash of
each field in the RangeQueryInfo struct. Note that we cannot do a range
query on state DB using hashed startKey and endKey to find phantom
reads.

Instead of hashed RangeQueryInfo per collection, i have added a string
field to hold a hash of entire private RangeQueryInfos of that collection.
During validation, we can compare the hash of entire private RangeQueryInfo
present in the tempDB with the hash present in the block. I thought of
doing the same for hashed write-set too as we may not store this value in stateDB.
However, it might be useful during validation phase (while comparing the
write set of Ti with read set of Ti+1 -- especially when the private
write-set is unavailable/not received).

If the peer is configured to move on without private write-set
(implicit: phantom read check is disabled), we must not use
RangeQueryInfo. Instead add the read <key, version> to private read
set and hashed read set.

I have also added PrivateCollectionsKVRWSet to hold private read-write
set of all collections. For clear differentiation on variable names, i
have used Reads, HashedReads, and PrivateReads (similarly for the Writes).

<!-- Provide a general summary of your changes in the Title above -->

## Description
<!-- Describe your changes in detail. -->

## Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Fixes #

## How Has This Been Tested?
<!-- If this PR does not contain a new test case, explain why. -->
<!-- Describe in detail how you tested your changes. -->

## Checklist:
<!-- To check a box, and an 'x': [x] -->
<!-- To uncheck box, add a space: [ ] -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff).
- [] I have either added documentation to cover my changes or this change requires no new documentation.
- [] I have either added unit tests to cover my changes or this change requires no new tests.
- [] I have run [golint](https://github.com/golang/lint) and have fixed valid warnings in code I have added or modified. This tool generates false positives so you may choose to ignore some warnings. The goal is clean, consistent, and readable code.

<!-- The continuous integration build process will run [make checks](https://github.com/hyperledger/fabric/blob/master/Makefile#L22) to confirm that tests pass and that code quality meets minimum standards. You may optionally run this locally as PRs will not be accepted until they pass. -->

Signed-off-by: Senthil <cendhu@gmail.com>

cendhu committed:
Add hashed and private read-write set to proto

This CR adds protobuf struct for hashed read/write set and private
read/write set.

To the existing KVRWSet, i have added a field to store hashed read write
set of all collections. For a private read, we can store the hash in key
field but keep the version as is. For private writes, we can store hash
of key and hash of value but store the IsDelete as is. Hence, reused
existing KVRead and KVWrite.

For the range query info, it does not make sense to store the hash of
each field in the RangeQueryInfo struct. Note that we cannot do a range
query on state DB using hashed startKey and endKey to find phantom
reads.

Instead of hashed RangeQueryInfo per collection, i have added a string
field to hold a hash of entire private RangeQueryInfos of that collection.
During validation, we can compare the hash of entire private RangeQueryInfo
present in the tempDB with the hash present in the block. I thought of
doing the same for hashed write-set too as we may not store this value in stateDB.
However, it might be useful during validation phase (while comparing the
write set of Ti with read set of Ti+1 -- especially when the private
write-set is unavailable/not received).

If the peer is configured to move on without private write-set
(implicit: phantom read check is disabled), we must not use
RangeQueryInfo. Instead add the read <key, version> to private read
set and hashed read set.

I have also added PrivateCollectionsKVRWSet to hold private read-write
set of all collections. For clear differentiation on variable names, i
have used Reads, HashedReads, and PrivateReads (similarly for the Writes).

manish-sethi reviewed (changes requested):
Hi Senthil - High level comment is that do not modify the existing unit tests in this case. Just add new unit tests. This is because, when we add 'HashedRWSet' to the message 'KVRWSet', it should not break the code that expects the older message structure of 'KVRWSet' (i.e., without 'HashedRWSet'). For instance, old blocks in the ledger may have the older message structure. Going a step ahead, I would suggest writing a special test as mentioned below.
Switch to the master branch, create a serialized message of type 'KVRWSet' and save the bytes in a temp file. Now in this branch check-in the file that has the serialized message from the master branch and commit this with the special unit-test that tries to deserialize the bytes from this file.

manish-sethi commented on protos/ledger/rwset/kvrwset/kv_rwset.proto:
```diff
@@ -21,11 +21,44 @@ option java_package = "org.hyperledger.fabric.protos.ledger.rwset.kvrwset";
+message HashedRWSet {
+    string collection = 1;
+    repeated KVRead hashed_reads = 2;
+    string hashed_range_query_info = 3;
```
See Comment at line 60. However, in either case, this should be declared as 'bytes' type. Am I missing something for this to be a 'string' type

manish-sethi commented on protos/ledger/rwset/kvrwset/kv_rwset.proto:
```diff
@@ -21,11 +21,44 @@ option java_package = "org.hyperledger.fabric.protos.ledger.rwset.kvrwset";
+message PrivateKVRWSet {
+    string collection = 1;
+    repeated KVRead private_reads = 2;
+    repeated RangeQueryInfo private_range_queries_info = 3;
```
I am not sure whether RangeQuery can be supported with the collection feature. Because, the validators who do not have access to a collection, cannot validate the RWSet. So, Ideally, collection and RangeQueryInfo fields should be declared 'either/or'. But for first cut, we can keep the `collection` in and `rangequeryInfo` out for simplicity.

manish-sethi reviewed (commented):
Hi Senthil - I have put some comments.

cendhu commented:
Hi Manish, I have added the requested tests and removed range queries from the private and hashed read write set. Later, we can add oneof with collection and range query. Please let me know if any changes are required.

manish-sethi commented on core/ledger/kvledger/txmgmt/rwsetutil/rwsetV1ProtoBytes.txt:
```diff
@@ -0,0 +1,21 @@
+K
```
Can you declare this file as a binary file. So that git does not interfere with the contents of this file while transferring across platforms

cendhu replied:
I am still trying to figure this out.

manish-sethi commented on protos/ledger/rwset/kvrwset/kv_rwset.proto:
```diff
@@ -21,11 +21,42 @@ option java_package = "org.hyperledger.fabric.protos.ledger.rwset.kvrwset";
+message HashedRWSet {
+    string collection = 1;
+    repeated KVRead hashed_reads = 2;
+    repeated KVWrite hashed_writes = 4;
```
3 instead of 4?

cendhu replied:
Done.

manish-sethi commented on core/ledger/kvledger/txmgmt/rwsetutil/rwset_proto_util_test.go:
```diff
@@ -63,6 +158,51 @@ func TestTxRWSetMarshalUnmarshal(t *testing.T) {
 	testutil.AssertNoError(t, txRwSet1.FromProtoBytes(protoBytes), "")
 	t.Logf("txRwSet=%s, txRwSet1=%s", spew.Sdump(txRwSet), spew.Sdump(txRwSet1))
 	testutil.AssertEquals(t, txRwSet1, txRwSet)
+	testutil.AssertNoError(t, err, "")
```
Forgot to remove this line?

cendhu replied:
Done.

manish-sethi commented on core/ledger/kvledger/txmgmt/rwsetutil/rwset_proto_util_test.go:
```diff
@@ -63,6 +158,51 @@ func TestTxRWSetMarshalUnmarshal(t *testing.T) {
 	testutil.AssertNoError(t, txRwSet1.FromProtoBytes(protoBytes), "")
 	t.Logf("txRwSet=%s, txRwSet1=%s", spew.Sdump(txRwSet), spew.Sdump(txRwSet1))
 	testutil.AssertEquals(t, txRwSet1, txRwSet)
```
swap the 2nd and 3rd arguments. I think that the 3rd argument to this function is 'expected value'

cendhu replied:
Done.

manish-sethi commented:
Hi Senthil, The changes look good to me. Just a couple of minor comments.
